File: Lib/glyphtracy/contour.py
import math
from dataclasses import dataclass

import numpy as np

from glyphtracy.node import AxisExtremaTag
from glyphtracy.utils import (
    axis_prominent_extrema_indices,
    dedupe_cyclic_indices,
    dedupe_open_indices,
    global_axis_extrema_axes,
    is_basically_a_line,
    normalized,
    prune_open_split_positions,
    smooth_signal,
    span_indices_closed,
    split_open_span_indices,
)


@dataclass
class Contour:
    contour_id: int
    points_rc: np.ndarray
    closed: bool = True

    @property
    def size(self) -> int:
        return int(self.points_rc.shape[0])

    def points_for_indices(self, indices: list[int]) -> np.ndarray:
        return self.points_rc[np.asarray(indices, dtype=np.int64)]

    def slice_before(self, contour_index: int, count: int) -> np.ndarray:
        n = self.size
        if n == 0:
            return np.empty((0, 2), dtype=np.float64)
        count = max(1, min(count, n))
        if self.closed:
            indices = [((contour_index - i) % n) for i in range(count, 0, -1)]
        else:
            start = max(0, contour_index - count)
            indices = list(range(start, contour_index))
        return self.points_for_indices(indices)

    def slice_after(self, contour_index: int, count: int) -> np.ndarray:
        n = self.size
        if n == 0:
            return np.empty((0, 2), dtype=np.float64)
        count = max(1, min(count, n))
        if self.closed:
            indices = [((contour_index + i) % n) for i in range(1, count + 1)]
        else:
            end = min(n, contour_index + count + 1)
            indices = list(range(contour_index + 1, end))
        return self.points_for_indices(indices)

    def find_inflection_indices(self, stride: int = 2) -> set[int]:
        n = self.size
        if n < 7:
            return set()

        signs: list[int] = []
        for i in range(n):
            cross = self.turning_cross_sign(i, stride=stride)
            if abs(cross) < 1e-5:
                signs.append(0)
            elif cross > 0:
                signs.append(1)
            else:
                signs.append(-1)

        inflections: set[int] = set()
        for i in range(n):
            prev_sign = signs[(i - 1) % n]
            cur_sign = signs[i]
            if prev_sign == 0 or cur_sign == 0:
                continue
            if prev_sign != cur_sign:
                inflections.add(i)
        return inflections

    def turning_cross_sign(self, i: int, stride: int = 1) -> float:
        n = self.size
        prev_i = (i - stride) % n if self.closed else max(0, i - stride)
        next_i = (i + stride) % n if self.closed else min(n - 1, i + stride)
        p_prev = self.points_rc[prev_i]
        p_cur = self.points_rc[i]
        p_next = self.points_rc[next_i]
        v_in = p_cur - p_prev
        v_out = p_next - p_cur
        return float(v_in[1] * v_out[0] - v_in[0] * v_out[1])

    def corner_angle(self, i: int, stride: int = 1) -> float:
        n = self.size
        if n < 3:
            return 0.0
        prev_i = (i - stride) % n if self.closed else max(0, i - stride)
        next_i = (i + stride) % n if self.closed else min(n - 1, i + stride)
        if prev_i == i or next_i == i:
            return 0.0
        p_prev = self.points_rc[prev_i]
        p_cur = self.points_rc[i]
        p_next = self.points_rc[next_i]
        v_in = normalized(p_cur - p_prev)
        v_out = normalized(p_next - p_cur)
        dot = float(np.dot(v_in, v_out))
        dot = max(-1.0, min(1.0, dot))
        return float(math.acos(dot))

    def iterative_axis_extrema_axes(
        self,
        *,
        min_index_gap: int = 3,
        min_span_points: int = 8,
        max_iterations: int = 8,
    ) -> dict[int, set[AxisExtremaTag]]:
        n = self.size
        if n < 3:
            return {}

        seed_map = global_axis_extrema_axes(
            self.points_rc,
            short_plateau_max_len=max(2, n // 80),
        )
        seed_indices = set(seed_map.keys())
        for axis in range(2):
            smoothed = smooth_signal(self.points_rc[:, axis], window=9, closed=True)
            axis_tag: AxisExtremaTag = "y" if axis == 0 else "x"
            axis_extrema = axis_prominent_extrema_indices(
                smoothed,
                radius=max(2, min(8, n // 30)),
                min_prominence=1.5,
                closed=True,
            )
            seed_indices.update(axis_extrema)
            for idx in axis_extrema:
                seed_map.setdefault(int(idx), set()).add(axis_tag)

        split_indices = dedupe_cyclic_indices(
            list(seed_indices), n, min_index_gap=max(2, min_index_gap)
        )
        discovered: dict[int, set[AxisExtremaTag]] = {
            int(idx): set(seed_map.get(int(idx), set())) for idx in split_indices
        }

        if len(split_indices) < 2:
            spans = [list(range(n))]
        else:
            spans: list[list[int]] = []
            for span_ix, start in enumerate(split_indices):
                end = split_indices[(span_ix + 1) % len(split_indices)]
                span = span_indices_closed(start, end, n)
                if len(span) >= 2:
                    spans.append(span)
            if not spans:
                spans = [list(range(n))]

        min_span_points = max(4, int(min_span_points))
        for _ in range(max_iterations):
            did_split = False
            next_spans: list[list[int]] = []

            for span_indices in spans:
                if len(span_indices) < max(2 * min_span_points, 6):
                    next_spans.append(span_indices)
                    continue

                subcontour = self.points_for_indices(span_indices)
                local_candidates = global_axis_extrema_axes(
                    subcontour,
                    short_plateau_max_len=max(2, min_index_gap),
                )
                for axis in range(2):
                    smoothed = smooth_signal(
                        subcontour[:, axis], window=5, closed=False
                    )
                    axis_tag: AxisExtremaTag = "y" if axis == 0 else "x"
                    axis_extrema = axis_prominent_extrema_indices(
                        smoothed,
                        radius=2,
                        min_prominence=0.75,
                        closed=False,
                    )
                    for local_ix in axis_extrema:
                        local_candidates.setdefault(int(local_ix), set()).add(axis_tag)

                split_positions = dedupe_open_indices(
                    list(local_candidates.keys()),
                    len(span_indices),
                    min_index_gap=max(2, min_index_gap - 1),
                )
                split_positions = prune_open_split_positions(
                    split_positions, len(span_indices), min_span_points=min_span_points
                )

                if not split_positions:
                    next_spans.append(span_indices)
                    continue

                did_split = True
                for local_ix in split_positions:
                    global_ix = span_indices[local_ix]
                    discovered.setdefault(int(global_ix), set()).update(
                        local_candidates.get(int(local_ix), set())
                    )
                next_spans.extend(
                    split_open_span_indices(span_indices, split_positions)
                )

            spans = next_spans
            if not did_split:
                break

        return discovered

    def iterative_axis_extrema_axes_on_open_span(
        self,
        span_indices: list[int],
        *,
        min_index_gap: int = 3,
        min_span_points: int = 8,
        max_iterations: int = 8,
    ) -> dict[int, set[AxisExtremaTag]]:
        if len(span_indices) < 3:
            return {}

        # Sharp-corner-bounded spans that are effectively straight should not
        # receive interior extrema nodes due to raster noise.
        span_points = self.points_for_indices(span_indices)
        if is_basically_a_line(span_points, pixel_tolerance=1.5):
            return {}

        local_all = list(range(len(span_indices)))
        discovered: dict[int, set[AxisExtremaTag]] = {}
        spans_local: list[list[int]] = [local_all]
        min_span_points = max(4, int(min_span_points))

        for _ in range(max_iterations):
            did_split = False
            next_local_spans: list[list[int]] = []

            for local_span in spans_local:
                if len(local_span) < max(2 * min_span_points, 6):
                    next_local_spans.append(local_span)
                    continue

                global_span = [span_indices[ix] for ix in local_span]
                subcontour = self.points_for_indices(global_span)

                local_candidates = global_axis_extrema_axes(
                    subcontour,
                    short_plateau_max_len=max(2, min_index_gap),
                )
                for axis in range(2):
                    smoothed = smooth_signal(
                        subcontour[:, axis], window=5, closed=False
                    )
                    axis_tag: AxisExtremaTag = "y" if axis == 0 else "x"
                    axis_extrema = axis_prominent_extrema_indices(
                        smoothed,
                        radius=2,
                        min_prominence=0.75,
                        closed=False,
                    )
                    for local_ix in axis_extrema:
                        local_candidates.setdefault(int(local_ix), set()).add(axis_tag)

                split_positions = dedupe_open_indices(
                    list(local_candidates.keys()),
                    len(local_span),
                    min_index_gap=max(2, min_index_gap - 1),
                )
                split_positions = prune_open_split_positions(
                    split_positions, len(local_span), min_span_points=min_span_points
                )

                if not split_positions:
                    next_local_spans.append(local_span)
                    continue

                did_split = True
                for local_ix in split_positions:
                    global_ix = span_indices[local_span[local_ix]]
                    discovered.setdefault(int(global_ix), set()).update(
                        local_candidates.get(int(local_ix), set())
                    )
                next_local_spans.extend(
                    split_open_span_indices(local_span, split_positions)
                )

            spans_local = next_local_spans
            if not did_split:
                break

        return discovered

File: glyphtracy-ts/src/contour.ts
import { AxisExtremaTag } from "./types.js";
import {
  axisProminentExtremaIndices,
  dedupeCyclicIndices,
  dedupeOpenIndices,
  globalAxisExtremaAxes,
  isBasicallyALine,
  normalized,
  pruneOpenSplitPositions,
  smoothSignal,
  spanIndicesClosed,
  splitOpenSpanIndices,
} from "./utils.js";

export class Contour {
  contourId: number;
  pointsRc: number[][];
  closed: boolean;

  constructor(contourId: number, pointsRc: number[][], closed = true) {
    this.contourId = contourId;
    this.pointsRc = pointsRc;
    this.closed = closed;
  }

  get size(): number {
    return this.pointsRc.length;
  }

  pointsForIndices(indices: number[]): number[][] {
    return indices.map((i) => this.pointsRc[i]);
  }

  cornerAngle(i: number, stride = 1): number {
    const n = this.size;
    if (n < 3) {
      return 0;
    }

    const prevI = this.closed ? (i - stride + n) % n : Math.max(0, i - stride);
    const nextI = this.closed ? (i + stride) % n : Math.min(n - 1, i + stride);
    if (prevI === i || nextI === i) {
      return 0;
    }

    const pPrev = this.pointsRc[prevI];
    const pCur = this.pointsRc[i];
    const pNext = this.pointsRc[nextI];

    const vin = normalized([pCur[0] - pPrev[0], pCur[1] - pPrev[1]]);
    const vout = normalized([pNext[0] - pCur[0], pNext[1] - pCur[1]]);
    const d = Math.max(-1, Math.min(1, vin[0] * vout[0] + vin[1] * vout[1]));
    return Math.acos(d);
  }

  iterativeAxisExtremaAxes(args: {
    minIndexGap?: number;
    minSpanPoints?: number;
    maxIterations?: number;
  }): Map<number, Set<AxisExtremaTag>> {
    const minIndexGap = args.minIndexGap ?? 3;
    const minSpanPoints = args.minSpanPoints ?? 8;
    const maxIterations = args.maxIterations ?? 8;

    const n = this.size;
    if (n < 3) {
      return new Map();
    }

    const seedMap = globalAxisExtremaAxes(
      this.pointsRc,
      Math.max(2, Math.floor(n / 80)),
    );
    const seedIndices = new Set<number>(seedMap.keys());

    for (let axis = 0; axis < 2; axis += 1) {
      const smoothed = smoothSignal(
        this.pointsRc.map((p) => p[axis]),
        9,
        true,
      );
      const axisTag: AxisExtremaTag = axis === 0 ? "y" : "x";
      const axisExtrema = axisProminentExtremaIndices(
        smoothed,
        Math.max(2, Math.min(8, Math.floor(n / 30))),
        1.5,
        true,
      );
      for (const idx of axisExtrema) {
        seedIndices.add(idx);
        if (!seedMap.has(idx)) {
          seedMap.set(idx, new Set());
        }
        seedMap.get(idx)?.add(axisTag);
      }
    }

    const splitIndices = dedupeCyclicIndices(
      [...seedIndices],
      n,
      Math.max(2, minIndexGap),
    );
    const discovered = new Map<number, Set<AxisExtremaTag>>();
    for (const idx of splitIndices) {
      discovered.set(idx, new Set(seedMap.get(idx) ?? []));
    }

    let spans: number[][];
    if (splitIndices.length < 2) {
      spans = [Array.from({ length: n }, (_, i) => i)];
    } else {
      spans = [];
      for (let i = 0; i < splitIndices.length; i += 1) {
        const start = splitIndices[i];
        const end = splitIndices[(i + 1) % splitIndices.length];
        const span = spanIndicesClosed(start, end, n);
        if (span.length >= 2) {
          spans.push(span);
        }
      }
      if (spans.length === 0) {
        spans = [Array.from({ length: n }, (_, i) => i)];
      }
    }

    const minSpan = Math.max(4, Math.floor(minSpanPoints));
    for (let it = 0; it < maxIterations; it += 1) {
      let didSplit = false;
      const nextSpans: number[][] = [];

      for (const spanIndices of spans) {
        if (spanIndices.length < Math.max(2 * minSpan, 6)) {
          nextSpans.push(spanIndices);
          continue;
        }

        const subcontour = this.pointsForIndices(spanIndices);
        const localCandidates = globalAxisExtremaAxes(
          subcontour,
          Math.max(2, minIndexGap),
        );

        for (let axis = 0; axis < 2; axis += 1) {
          const smoothed = smoothSignal(
            subcontour.map((p) => p[axis]),
            5,
            false,
          );
          const axisTag: AxisExtremaTag = axis === 0 ? "y" : "x";
          const axisExtrema = axisProminentExtremaIndices(
            smoothed,
            2,
            0.75,
            false,
          );
          for (const localIx of axisExtrema) {
            if (!localCandidates.has(localIx)) {
              localCandidates.set(localIx, new Set());
            }
            localCandidates.get(localIx)?.add(axisTag);
          }
        }

        let splitPositions = dedupeOpenIndices(
          [...localCandidates.keys()],
          spanIndices.length,
          Math.max(2, minIndexGap - 1),
        );
        splitPositions = pruneOpenSplitPositions(
          splitPositions,
          spanIndices.length,
          minSpan,
        );

        if (splitPositions.length === 0) {
          nextSpans.push(spanIndices);
          continue;
        }

        didSplit = true;
        for (const localIx of splitPositions) {
          const globalIx = spanIndices[localIx];
          if (!discovered.has(globalIx)) {
            discovered.set(globalIx, new Set());
          }
          const tags =
            localCandidates.get(localIx) ?? new Set<AxisExtremaTag>();
          for (const tag of tags) {
            discovered.get(globalIx)?.add(tag);
          }
        }

        nextSpans.push(...splitOpenSpanIndices(spanIndices, splitPositions));
      }

      spans = nextSpans;
      if (!didSplit) {
        break;
      }
    }

    return discovered;
  }

  iterativeAxisExtremaAxesOnOpenSpan(
    spanIndices: number[],
    args: {
      minIndexGap?: number;
      minSpanPoints?: number;
      maxIterations?: number;
    },
  ): Map<number, Set<AxisExtremaTag>> {
    const minIndexGap = args.minIndexGap ?? 3;
    const minSpanPoints = args.minSpanPoints ?? 8;
    const maxIterations = args.maxIterations ?? 8;

    if (spanIndices.length < 3) {
      return new Map();
    }

    const spanPoints = this.pointsForIndices(spanIndices);
    if (isBasicallyALine(spanPoints, 1.5)) {
      return new Map();
    }

    const localAll = Array.from({ length: spanIndices.length }, (_, i) => i);
    const discovered = new Map<number, Set<AxisExtremaTag>>();
    let spansLocal: number[][] = [localAll];
    const minSpan = Math.max(4, Math.floor(minSpanPoints));

    for (let it = 0; it < maxIterations; it += 1) {
      let didSplit = false;
      const nextLocalSpans: number[][] = [];

      for (const localSpan of spansLocal) {
        if (localSpan.length < Math.max(2 * minSpan, 6)) {
          nextLocalSpans.push(localSpan);
          continue;
        }

        const globalSpan = localSpan.map((ix) => spanIndices[ix]);
        const subcontour = this.pointsForIndices(globalSpan);
        const localCandidates = globalAxisExtremaAxes(
          subcontour,
          Math.max(2, minIndexGap),
        );

        for (let axis = 0; axis < 2; axis += 1) {
          const smoothed = smoothSignal(
            subcontour.map((p) => p[axis]),
            5,
            false,
          );
          const axisTag: AxisExtremaTag = axis === 0 ? "y" : "x";
          const axisExtrema = axisProminentExtremaIndices(
            smoothed,
            2,
            0.75,
            false,
          );
          for (const localIx of axisExtrema) {
            if (!localCandidates.has(localIx)) {
              localCandidates.set(localIx, new Set());
            }
            localCandidates.get(localIx)?.add(axisTag);
          }
        }

        let splitPositions = dedupeOpenIndices(
          [...localCandidates.keys()],
          localSpan.length,
          Math.max(2, minIndexGap - 1),
        );
        splitPositions = pruneOpenSplitPositions(
          splitPositions,
          localSpan.length,
          minSpan,
        );

        if (splitPositions.length === 0) {
          nextLocalSpans.push(localSpan);
          continue;
        }

        didSplit = true;
        for (const localIx of splitPositions) {
          const globalIx = spanIndices[localSpan[localIx]];
          if (!discovered.has(globalIx)) {
            discovered.set(globalIx, new Set());
          }
          const tags =
            localCandidates.get(localIx) ?? new Set<AxisExtremaTag>();
          for (const tag of tags) {
            discovered.get(globalIx)?.add(tag);
          }
        }

        nextLocalSpans.push(...splitOpenSpanIndices(localSpan, splitPositions));
      }

      spansLocal = nextLocalSpans;
      if (!didSplit) {
        break;
      }
    }

    return discovered;
  }
}
